src/orange/construct/detail/NodeReplacementInserter.hh

```diff
@@ -0,0 +1,142 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/construct/detail/NodeReplacementInserter.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <variant>
+#include <vector>
+
+#include "../CsgTree.hh"
+#include "../CsgTypes.hh"
+
+namespace celeritas
+{
+namespace csg
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Add a node ID and its "replaced" value.
+ *
+ * This implementation detail of the "replace down" algorithm adds daughters
+ * to a queue of nodes to visit, along with their replacement values.
+ */
+class NodeReplacementInserter
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using VecNode = std::vector<std::pair<NodeId, Node>>;
+    //!@}
+
+  public:
+    // Construct with references to the stack and the replacement value
+    inline NodeReplacementInserter(VecNode* stack, Node const& repl);
+
+    // If replacing a queued node with a boolean, ours should match
+    inline void operator()(True const&);
+    inline void operator()(False const&);
+
+    // Surfaces cannot be simplified further
+    void operator()(Surface const&) {}
+
+    // Simplify node types that reference other nodes
+    inline void operator()(Aliased const& n);
+    inline void operator()(Negated const& n);
+    inline void operator()(Joined const& n);
+
+  private:
+    VecNode* stack_;
+    Node const& repl_;
+    Node negated_;
+    OperatorToken join_token_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with references to the stack and the replacement value.
```

Comment:
pointer

Reply:
In the doc strings I think pointer and reference can be effectively interchangeable. The type itself is clear from the code, the point of the documentation is to emphasize that you're referencing memory outside the class itself.

Reply:
I agree that you shouldn't have to mention the type in a doc string, but if you are going to mention a type, it should be the correct type.